### DISCUSSION TITLE
Generic Data Index Bundle doesn't get all elements from Pimcore

### DISCUSSION THREAD

**Discussion by khenow1978** [pimcore/pimcore#17222] [General]
Hi.. after updating to v11.2.6 and setting up an AWS Opensearch Server I've started to install the generic-index-bundle and initialized the index process of our portal Engine Portal. Unfortunately it sort of misses a lot of entries. Right now, this is what a user sees in our portal:

![image](https://github.com/pimcore/pimcore/assets/92732594/a66b39dd-991e-4541-bf3e-bc867627246d)

So some pdf files got indexed, some didn't and those aren't reachable within the portal. I tried several options, e. g. I restarted the process with "bin/console generic-data-index:update:index -r" I have updated the index several times but with no visible effect.

I've noticed that sometimes the messenger:stats shows a failed entry and when I tried to retry those entries, the following appeared:

![image](https://github.com/pimcore/pimcore/assets/92732594/06941811-43fd-44a2-a7f3-71c42c48361a)

or this message pops up in between:

![image](https://github.com/pimcore/pimcore/assets/92732594/2300dbf1-8f00-4dc7-97ed-cace594219b0)

The total number of items in the queue changes from time to time.. sometimes it's about 7k, then it goes up to 9 or 10k:

![image](https://github.com/pimcore/pimcore/assets/92732594/a20e6994-f551-43a5-8e5e-8b42d389d61f)

But I can say that a lot of stuff is currently missing in the Portal. The messenger stats are 0:

![image](https://github.com/pimcore/pimcore/assets/92732594/e6ce61d0-dccf-4a37-873f-c4a1ed3faed7)

Any idea what's going on here??

**Comment by markus-moser** [comment]
@khenow1978 Seems that there is some problem with the thumbnail generation.

* Did you try it out with generic data index bundle 1.1.0? There something improved in the error handling. When the thumbnail generation fails now the document should be still indexed (but with broken thumbnail).

* Do you store the assets in S3? Maybe there sometimes something failes when generating the thumbnails?

**Reply by khenow1978** [reply]
Hi.. I just tried, but as I use platform version 2024.1 the version is fixed to 1.0.99:

![image](https://github.com/pimcore/pimcore/assets/92732594/0e9f133c-677f-4e9f-8f71-cc3c135046af)

The assets are stored on the Pimcore Server in the public/var/ folder.

**Reply by markus-moser** [reply]
Hi,
platform version 2024.2 with the new version will be released on Tuesday next week.

But still it seems that there are some problems with the thumbnail generation. Do they get generated without a problem outside of the generic data index (so for example in the pimcore backend)?

**Reply by khenow1978** [reply]
tldr;
After removing all temp files and restarting the whole process the queue got shorter but is still not being completed.


Unfortunately I don't know how to check whether the thumbnails have been created or not. I did the next best thing: Checked the filesystem in the tmp folder for the files and compared those to the opensearch indexes.. there you'll find the custom_fields.PortalEngineBundle.systemfields.thumbnail link. I realized that within the file system Pimcore didn't create all "portal-egine_element-teaser" files.

So in my test environment on localhost I started removing all tmp files and cleared the cache. First it didn't do shit.. really nothing. I could create the indexes from scratch.. nothing happened. In the log files I could see that opensearch was opening a "stats" site.. but that's all. 

After a few minutes I saw that Pimcore was still working on creating all the thumbnails in the tmp folders.. but just the "system-treepreview".. not the "element-teaser" for the portal engine. That came later and then it also started to send indexes to Opensearch.

Right now I'm down to 5500 remaining elements.. I see stuff in the Portal Engine but there's still a lot missing. In the log file there are no new pimcore.opensearch entries.. it looks as if Pimcore thinks the job is completely finished.. but it isn't.

**Reply by khenow1978** [reply]
I did some further testing, identified some missing objects and checked the transport messages.. and now I got some error messages:


Fehler 1 - Objekt:
15:17:45 INFO      [pimcore.opensearch] Request Success: ["method" => "POST","uri" => "https://thisismyurl.com:443/_bulk?refresh=wait_for","port" => 443,"headers" => ["Host" => ["thisismyurl.amazonaws.com"],"Content-Type" => ["application/json"],"Accept" => ["application/json"],"User-Agent" => ["opensearch-php/2.3.0 (Linux 5.10.200+1-ph; PHP 8.2.18)"]],"HTTP code" => 200,"duration" => 0.017292]
15:17:45 WARNING   [pimcore] handleIndexQueueEntry failed! Error: OpenSearch bulk produced errors: {"took":1,"errors":true,"items":[{"update":{"_index":"portal_live_artikel-odd","_id":"6327","status":400,"error":{"type":"mapper_parsing_exception","reason":"failed to parse field [standard_fields.merkmaleBrick.Entwaesserungsdetails.wasserAblaufTable] of type [text] in document with id '6327'. Preview of field's value: '{value=0,7, height=5}'","caused_by":{"type":"illegal_state_exception","reason":"Can't get text on a START_OBJECT at 1:6776"}}}}]}
15:17:45 INFO      [messenger] Message Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage handled by Pimcore\Bundle\GenericDataIndexBundle\MessageHandler\IndexUpdateQueueHandler::__invoke ["class" => "Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage","handler" => "Pimcore\Bundle\GenericDataIndexBundle\MessageHandler\IndexUpdateQueueHandler::__invoke"]
15:17:45 INFO      [messenger] Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage was handled successfully (acknowledging to transport). ["class" => "Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage"]


Fehler Asset:
15:22:48 INFO      [pimcore.opensearch] Request Success: ["method" => "POST","uri" => "https://thisimyurl.com:443/_bulk?refresh=wait_for","port" => 443,"headers" => ["Host" => ["thissimyurl.com"],"Content-Type" => ["application/json"],"Accept" => ["application/json"],"User-Agent" => ["opensearch-php/2.3.0 (Linux 5.10.200+1-ph; PHP 8.2.18)"]],"HTTP code" => 200,"duration" => 0.055847]
15:22:48 WARNING   [pimcore] handleIndexQueueEntry failed! Error: OpenSearch bulk produced errors: {"took":4,"errors":true,"items":[{"update":{"_index":"portal_live_asset-odd","_id":"600","status":400,"error":{"type":"illegal_argument_exception","reason":"Document contains at least one immense term in field=\"system_fields.text\" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[65, 66, 83, 32, 66, 97, 115, 105, 99, 10, 10, 77, 111, 65, 32, 50, 49, 54, 45, 48, 49, 10, 10, 80, 83, 45, 66, 65, 83, 73]...', original message: bytes can be at most 32766 in length; got 117226","caused_by":{"type":"max_bytes_length_exceeded_exception","reason":"bytes can be at most 32766 in length; got 117226"}}}}]}
15:22:48 INFO      [messenger] Message Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage handled by Pimcore\Bundle\GenericDataIndexBundle\MessageHandler\IndexUpdateQueueHandler::__invoke ["class" => "Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage","handler" => "Pimcore\Bundle\GenericDataIndexBundle\MessageHandler\IndexUpdateQueueHandler::__invoke"]
15:22:48 INFO      [messenger] Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage was handled successfully (acknowledging to transport). ["class" => "Pimcore\Bundle\GenericDataIndexBundle\Message\IndexUpdateQueueMessage"]

However that doesn't really give me any explanation on how to resolve this. Any idea from your side? Is this something based on our data model or is it a more general opensearch issue??

**Reply by markus-moser** [reply]
This seems to be a bug - I created an issue and will fix it:

https://github.com/pimcore/generic-data-index-bundle/issues/175

**Reply by khenow1978** [reply]
Both cases? In the objects it looks as if there is a table with several rows it's declining the update as well but I'm not sure if this is due to its length or because the expected data type doesn't fit to a table?!

**Reply by markus-moser** [reply]
I will take a look at the table topic too. So is this a field with type "table" or "structured table"?

**Reply by khenow1978** [reply]
It's just a simple table. There are 2 tables set up in the data model. As long as they are not filled with data the object gets indexed without any errors. After I added a row the index stops with the above mentioned error message.

**Reply by markus-moser** [reply]
@khenow1978 The asset case is fixed in https://github.com/pimcore/generic-data-index-bundle/pull/176. For the case with the data object tables I'm currently not possible to reproduce it. Could you please send a screenshot how the field is configured in the class editor and which data you try to save in the object into the field?

**Reply by khenow1978** [reply]
![image](https://github.com/pimcore/pimcore/assets/92732594/60e38d0b-d08e-47a2-8c31-2c6da4ed05ed)

![image](https://github.com/pimcore/pimcore/assets/92732594/2ad4b6d3-f881-4f36-9428-be01b6b9ce36)

Above you'll find the screenshot from the class definitions tab. I have also reproduced it with a new class. I just created a string field and a table. This is how it looks like:

![image](https://github.com/pimcore/pimcore/assets/92732594/b2512645-a55a-45b8-a697-82e8304bd5e2)

As long as the table remains empty the object is indexed immediately. When I create a row or more the index update returns this:

![image](https://github.com/pimcore/pimcore/assets/92732594/cdb13aec-1888-402d-8bdb-72d9fd55090a)

This is the content Pimcore tries to store in there:  '{baseUnit=werefe, baseValue=55, targetUnit=sdff, factor=wddfe}'

This is how it looks like on the OpenSearch Console:

![image](https://github.com/pimcore/pimcore/assets/92732594/89129585-7fdb-40b3-9c99-e751d3303214)

It looks as if the tblFieldValues are expecting an array?

This happens on test and on prod respectively.

**Reply by markus-moser** [reply]
Thanks a lot - I guess the configured column configuration destroys it. It currently seems to work without defined column configuration only. I will check that.

**Reply by khenow1978** [reply]
I just did some manual testing and tried to update the index using the API of Opensearch:
![image](https://github.com/pimcore/pimcore/assets/92732594/125f9201-30ed-4838-a0ee-2e9586c468e4)

And this works.. I could add it as array with [] or without the brackets.. both ways are working.. but not when Pimcore tries to do the same.. :/

**Reply by markus-moser** [reply]
No pimcore will not try to index it like this. You encoded the JSON object as string -> pimcore will try to index it as JSON. We need to adjust the opensearch mapping to be able to handle the fixed column layout. I will fix this!

**Reply by khenow1978** [reply]
Hi.. I have just updated to version 1.1.1 and realized that there's a new error in the log files:

pimcore.ERROR: Update indexQueue in database-table generic_data_index_queue failed! Error: An exception occurred while executing a query: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'ON DUPLICATE KEY         UPDATE             operation = VALUES(operation),      ' at line 4

As far as I remember that wasn't an error in the last version.. :/

By the way.. the PDF length issue still remains..

**Reply by markus-moser** [reply]
Hi,
did you execute `bin/console generic-data-index:update:index`and restart your messenger workers? Then the PDF length issue should definitely be fixed.

**Reply by khenow1978** [reply]
Just updating the indexes wasn't possible:

 [pimcore.opensearch] Request Failure: ["method" => "PUT","uri" => "https://url.com:443/portal_local_asset-odd/_mapping","port" => 443,"headers" => ["Host" => ["urlom"],"Content-Type" => ["application/json"],"Accept" => ["application/json"],"User-Agent" => ["opensearch-php/2.3.0 (Linux 5.15.153.1-microsoft-standard-WSL2; PHP 8.2.12)"]],"HTTP code" => 400,"duration" => 0.017828,"error" => "{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"mapper [system_fields.elementType] cannot be changed from type [text] to [keyword]"}],"type":"illegal_argument_exception","reason":"mapper [system_fields.elementType] cannot be changed from type [text] to [keyword]"},"status":400}"]

So I had to recreate the indexes from scratch.. at the moment I'm testing it in my test environment and it's still running.

**Reply by markus-moser** [reply]
Yes the mapping update can fail but then the generic data index will create a new index and automatically switches it. So re-creation of the index should not be needed.

**Reply by khenow1978** [reply]
Sorry to bother you again but the behaviour is still pretty odd.. yesterday I started the index process with version 1.1.1 but til today not every pdf File get indexed. this is how it looks like in the portal engine:

![image](https://github.com/pimcore/pimcore/assets/92732594/ab2231d5-65a8-449b-b124-2ff523673cde)

Some files are still missing because you can't open them.. BUT when I open such a file in the pimcore backend and just save it (without doing anything else.. open the file, click on "Save & publish" it suddently gets indexed:

![image](https://github.com/pimcore/pimcore/assets/92732594/94a2f80e-1785-42a6-ba6f-59da75461d4a)

Why does this happend and do I have to open every single file in order to get it indexed???

I have also compoared the amount of files in the backend with the amount in the portal engine:

Product manuals: 221 in the backend, 221 in the portal engine
Product drawings PDF files: 799 in the backend, 799 in the portal engine
Product drawings STP files: 278 in the backend, 278 in the portal engine
Certificates PDF Files: 371 in the backend, only 131 in the portal engine

I don't really understand what's going on here.. it seem as if the index process stopped in between but I can't find any error message in the log files or any hint why the certificates can't get indexed completely.. :/

**Reply by khenow1978** [reply]
@markus-moser : Any news on this? Even with the latest version I'm missing several hundreds PDF Files and a lot of products and there's still no real solution on how to resolve this by myself. 

That's why I have to wait for an update with a fix.
BR
Karsten

**Reply by markus-moser** [reply]
Hi Karsten,
sorry, I was on vacation. I will fix the issue with the table data type very soon. I'm currently not able to reproduce the PDF topic. Don't know how critical the content of the effected PDFs is but would it be possible that you send one example PDF to me which currently does not work or only works when you save it again in the backend?

**Reply by khenow1978** [reply]
Hi @markus-moser .. vacation.. is it THAT time again? ;)

I would love to send you a pdf that's not working .. the only problem is: the PDF are working.. if I trigger the index manually. As described above, if I open the file it's indexed immediately. But that's not done automatically.. it looks as if the process got stuck somewhere in between. What I noticed: All pdf files which are not part of the Index yet show this message when I open them in Pimcore:

![image](https://github.com/user-attachments/assets/ee6e0d45-d8b4-4d5d-88c1-2d526ced53fa)

When this process is finished they are automatically processed to the index. First, I thought I have to click on "Save & publish" but that's not needed. I just have to wait for scanning to be finished. So if I open the 200 missing files I would complete the index.. but that's something I really don't wanna do for obvious reasons..

Right now the queue window in the Pimcore admin still got 2100 open elements although the messenger queue tells me that there are no open items. Very weird..

**Reply by markus-moser** [reply]
Thanks for this additional hint - I will try to find out what's the problem.

**Reply by markus-moser** [reply]
@khenow1978 I still think that it might have something to do with the concrete asset. Could you please send an example PDF where this happend?

**Reply by khenow1978** [reply]
[ABS Basic - DEKRA Prüfbericht.pdf](https://github.com/user-attachments/files/16235353/ABS.Basic.-.DEKRA.Prufbericht.pdf)

This one hasn't been indexed.. now it has when I opened it ^^

**Reply by markus-moser** [reply]
Really strange - this PDF works for me too without any problems  - also while the PDF preview is not already available and this message would appear:

![image](https://github.com/user-attachments/assets/3ae69a64-595e-410a-9db2-e7c81f1efe0d)

Could you please take a look into the `generic_data_index_queue` table if there is an entry for one of the other PDFs which is currently missing?

Could you please also execute the following SQL statement and tell me if there is a result?
`select * from tmp_store where id = 'generic_data_index_queue_messages_dispatch_pending';`

**Reply by khenow1978** [reply]
The table contains all the ids which haven't been sent to the index yet:

![image](https://github.com/user-attachments/assets/933e27b6-99b8-4348-9854-72e5360b25fe)

I also opened some and stored them.. they were immediately indexed and removed from the table. However somehow Pimcore isn't triggered to work on those entries on its own.. as you see on the top of the screen: The table contains 2086 elements from different classes.

The other statement doesn't show any results. The tmp_store itself just holds 10 records at the moment.. none of them are related to the data_index_queue..

**Reply by markus-moser** [reply]
Could you please execute the following SQL statement?

`update generic_data_index_queue set dispatched=0 where elementType='asset'`

And then run `bin/console generic-data-index:update:index`

Do the assets then get indexed?

**Reply by markus-moser** [reply]
I think that for some reason those assets failed in previous queue message handling tries. That they have the dispatched column filled with a value means that messages for those items where already dispatched. If you do not have them in your message queue this means that the execution of those messages failed. Unfortunately there is a bad try/catch block which prevents the messages going to the failed items queue in those cases. I fixed this in this PR:

https://github.com/pimcore/generic-data-index-bundle/pull/178

So now failed items will go to the failed items queue which would make it possible to retry them later and either see what's the problem or get them finally processed (in case there was a temporary problem like that Opensearch was down or whatever). 

Resetting the dispatched column will trigger a new message when `bin/console generic-data-index:update:index` gets executed and therefore retry the items.

**Reply by markus-moser** [reply]
The table data type should be fixed with this PR:

https://github.com/pimcore/generic-data-index-bundle/pull/179

Tomorrow we will do the release of the 1.1.2 version then you are able to test it. Hopefully then all problems are solved.

**Reply by khenow1978** [reply]
Good Morning,

I just tested it. Before, the queue contained different object classes but now that's down to 5000 objects of the type "Artikel":

![image](https://github.com/user-attachments/assets/e6583ee6-90ce-4eda-acaf-3f9b8170e926)

And now, nothing happens. This is what messenger:stats tells me:

![image](https://github.com/user-attachments/assets/3e3379a6-c1b8-4343-9cfa-d58114d6960a)

I tried to retry those messages with the command mentioned in the documentation and this is what I get:

![image](https://github.com/user-attachments/assets/c4dae04a-eaad-4204-a89e-d75863db1384)

![image](https://github.com/user-attachments/assets/68108073-0e49-4136-8f34-aa3adaf01875)


I'm not sure why this happens or what the root cause may be.. there are assets though where Pimcore can't create thumbnails, like .dwg or .stp files for example. But would it cause the queue to break completely?

I will test the update as soon as it has been released.. :)

**Reply by markus-moser** [reply]
Hi Karsten,

I created the release. Please update the generic data index to 1.1.2 and also the portal engine to 4.1.2 and try it again, thanks!

**Reply by khenow1978** [reply]
It works! :) At least to a degree of 99,9%. I found 2 errors:

1) That's on our side.. I found a pdf file which is corrupt. Neither Pimcore nor my computer is able to open it "the pdf file is invalid". When the dispatch queue reaches this document it stops immediately. After I removed this id from the generic_index_queue table and set the dispatchdate to 0 all elements were processed. I've asked a colleague of mine to check the file and provide a new version.

2) That's a weird one because it looks as if the index is incompatible with some data provided by the blackbit data director which we are using. Although every element is being processed I see a lot of failed messages like this: 

![image](https://github.com/user-attachments/assets/309f222d-e1f4-4c9f-9779-399338456b00)

I may have to check this on a longterm basis if the failed transports number just keeps increasing or if Pimcore somehow ignores these elements.

But anyway, many thanks for your help and I bet that this is the longest discussions thread ever ;)

BR,
Karsten

**Reply by markus-moser** [reply]
Hi Karsten,

Great - yes the second issue seem to be not related to the generic data index (it's a different queue).

Many thanks for your valuable input, it helped a lot! 
It's not that easy to find out what's the problem with this kind of "discussion based debugging", so I'd say we were able to manage it in a reasonable amount of comments ;)

Best,
Markus
